Zend/zend_execute_API.c

```diff
@@ -1323,6 +1336,57 @@ static void zend_timeout_handler(int dummy) /* {{{ */
 /* }}} */
 #endif
 
+#ifndef ZEND_WIN32
+static void zend_wall_timeout_handler(int dummy) /* {{{ */
+{
+#ifndef ZTS
+    if (EG(timed_out)) {
+		/* Die on hard timeout */
+		const char *error_filename = NULL;
+		uint32_t error_lineno = 0;
+		char log_buffer[2048];
+		int output_len = 0;
+
+		if (zend_is_compiling()) {
+			error_filename = ZSTR_VAL(zend_get_compiled_filename());
+			error_lineno = zend_get_compiled_lineno();
+		} else if (zend_is_executing()) {
+			error_filename = zend_get_executed_filename();
+			if (error_filename[0] == '[') { /* [no active file] */
+				error_filename = NULL;
+				error_lineno = 0;
+			} else {
+				error_lineno = zend_get_executed_lineno();
+			}
+		}
+		if (!error_filename) {
+			error_filename = "Unknown";
+		}
+
+		output_len = snprintf(log_buffer, sizeof(log_buffer), "\nFatal error: Maximum execution wall-time of " ZEND_LONG_FMT "+" ZEND_LONG_FMT " seconds exceeded (terminated) in %s on line %d\n", EG(wall_timeout_seconds), EG(hard_timeout), error_filename, error_lineno);
+		if (output_len > 0) {
+			zend_quiet_write(2, log_buffer, MIN(output_len, sizeof(log_buffer)));
+		}
+		_exit(124);
+    }
+#endif
+
+	if (zend_on_timeout) {
+		zend_on_timeout(EG(wall_timeout_seconds));
+	}
+
+	EG(timed_out) = 2;
+	EG(vm_interrupt) = 1;
+
+#ifndef ZTS
+	if (EG(hard_timeout)) {
+		zend_set_wall_timeout_ex(EG(hard_timeout), 1);
```

Comment:
Regarding https://externals.io/message/112492#112783, my understanding is that the following happens:
- when `max_execution_wall_time` is reached, the necessary globals are set (`timed_out` and `vm_interrupt`)
- the timer is restarted with the `hard_timeout` when it has a value other than `0` (BTW: shouldn't we disallow negative values?)
- if the timeout is reached again, the process gets aborted

What is not clear for me, why timeouts inside internal functions can cause a process abort immediately, given their individual timeout is >= `hard_timeout`? What am I missing?

Another thing I've just become unsure about is whether it is really a good idea to restart the real-time timer instead of the original CPU-time one:
 ```suggestion
		zend_set_timeout_ex(EG(hard_timeout), 1);
```